---
sort: 2
title: "OTUanalysis"
author: "XiaZhao"
date: "2021-04-13"
output: 
  md_document: 
    df_print: kable
    number_sections: yes
    toc: yes
---

# OTUanalysis

=========

## Requires

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
<details>
<summary>
<font size=4>Requires</font>
</summary>
```{r tidy=TRUE,results='hold',message=FALSE}
library(tidyverse)
library(ggthemes)
library(ggsci)
library(ggpubr)
library(survminer)
library(survival)
library(survivalROC)
library(reshape2)
library(data.table)
library(ggExtra)
library(cowplot)
library(pheatmap)
library(scico)
library(colorspace)
library(RColorBrewer)
library(lubridate)
library(tableone)
library(kableExtra)
library(BiodiversityR)
library(reactable)
source("../R_function/colors.R")
source("../R_function/surv_plot.R")
source("../R_function/OTUanalysis.R")
theme_set(theme_cowplot())
"%ni%"<-Negate("%in%")
options(stringsAsFactors = F)
```
</details>

## Summary


### OTU（Operational Taxonomic Units）
In 16S metagenomics approaches, OTUs are cluster of similar sequence variants of the 16S rDNA marker gene sequence. Each of these cluster is intended to represent a taxonomic unit of a bacteria species or genus depending on the sequence similarity threshold. Typically, OTU cluster are defined by a 97% identity threshold of the 16S gene sequences to distinguish bacteria at the genus level.

#### Limited taxonomic resolution 
 
OTU resolution depends on the 16S approach which has some limits in distinguishing at the species level

### Code

## Rank-Abundance

A rank abundance curve or Whittaker plot is a chart used by ecologists to display relative species abundance, a component of biodiversity. It can also be used to visualize species richness and species evenness. It overcomes the shortcomings of biodiversity indices that cannot display the relative role different variables played in their calculation.

### Code

```{r warning=F}
otu<-fread("../Data/Data/OTUtable_ori.csv",data.table = F)
meta<-fread("../Data/Data/meta.csv",data.table = F)
otu<-t(data.frame(row.names = otu$OTU,otu[,-1]))
otu_relative <- otu / rowSums(otu)
rank_dat <- data.frame()
for (i in rownames(otu_relative)) {
  rank_dat_i <- data.frame(rankabundance(subset(otu_relative, 
                                                rownames(otu_relative) == i), 
                                         digits = 6))[1:2]
  rank_dat_i$sample <- i
  rank_dat <- rbind(rank_dat, rank_dat_i)
}
rank_dat <- subset(rank_dat, abundance != 0)
colnames(rank_dat)[3]="Samples"
rank_dat<-merge(rank_dat,meta,by="Samples")
p<-ggplot(rank_dat, aes(rank, log(abundance, 10), color = Site)) +
  geom_line(size=0.2) +
  scale_colour_manual(limits = c('Saliva','Stool'), values = c('darkblue','orange')) +
  labs(x = 'OTUs rank', y = 'Relative adundance (%)', color = NULL) +
  theme(panel.grid = element_blank(), panel.background = element_rect(fill = 'transparent', color = 'black'), legend.key = element_rect(fill = 'transparent')) +
  scale_y_continuous(breaks = 0:-5, labels = c('100', '10', '1', '0.1', '0.01', '0.001'), limits = c(-5, 0))
```

### Figures

```{r out.width = "50%",fig.height=4,fig.width=6, dpi=600,fig.align='center'}
p
```

### Effective sequences

```{r out.width = "50%",fig.height=4,fig.width=6, dpi=600,fig.align='center'}
ggplot(meta,aes(reorder(Samples,Sequences),Sequences,fill=Site))+
  geom_col()+
  scale_fill_jama()+
  facet_grid(~Site,scales = "free",space = "free")+
  theme(axis.text.x = element_blank())+
  xlab("Samples")
```

### Taxon annotation

```{r}
OTUstool<-fread("../Data/Data/OTUtabale_regaStool.csv",data.table = F)
OTUsaliva<-fread("../Data/Data/OTUtabale_regaSaliva.csv",data.table = F)
message("Taxon number on stool samples")
for (i in 1:7) {

  print(paste0(colnames(OTUstool)[i],":",
                 length(levels(factor(OTUstool[,i])))))
}
message("Taxon number on saliva samples")
for (i in 1:7) {
  
  print(paste0(colnames(OTUsaliva)[i],":",
                 length(levels(factor(OTUsaliva[,i])))))
}

```

# Pan/Core Microbiome
Pan/Core物种分析用于描述随着样本量增加物种总量和核心物种量变化的情况，在微生物多样性和群落研究中，被广泛用于判断样本量是否充足以及评估环境中总物种丰富度( species richness)和核心物种数。Pan/Core物种分析可以在各分类学水平上进行，包括domain（域），kingdom（界），phylum（门），class（纲），order（目），family（科），genus（属），species（种），OTU等。

## Pan Microbiome
Pan物种，即泛物种，是所有样本包含的物种总和，用于观测随着样本数目增加，总物种数目的增加情况。

## Core Microbiome

Core物种，即核心物种，是所有样本共有物种数目，用于观测随着样本数目增加，共有物种数目的减少情况。

One of the aims of the Human Microbiome Project when established in 2007 was to identify a human ‘core microbiome’, defined as a group of microbial taxa or genes that are shared by all or most humans (Hamady & Knight, 2009; Turnbaugh et al., 2007). These pioneering studies found that a universal taxonomic core rarely exists across groups of humans, even at the scale of the family unit (Yatsunenko et al., 2012), yet most shared the same set of core microbial genes. This pattern appears to be similar for most host species studied to date, with variable microbial composition often underpinned by similar gene context across individuals (Burke, Steinberg, Rusch, Kjelleberg, & Thomas, 2011; Louca et al., 2018). Nevertheless, most ecologists aim to understand host–microbe interactions at the organismal level, accounting for microbe functional traits where possible; therefore, within this field, the core microbiome has been largely applied to taxonomically defined microbial communities with an aim to identify groups of microbes that are particularly widespread across the host population (e.g. Ainsworth et al., 2015; Grieneisen, Livermore, Alberts, Tung, & Archie, 2017; Hernandez‐Agreda, Gates, & Ainsworth, 2017; Kembel et al., 2014; Lundberg et al., 2012; Muletz Wolz, Yarwood, Campbell Grant, Fleischer, & Lips, 2018).

ref:Applying the core microbiome to understand host–microbe systems.J Anim Ecol. 2020;89:1549–1558 (2020). https://doi.org/10.1111/1365-2656.13229

ref: Turnbaugh, P., Hamady, M., Yatsunenko, T. et al. A core gut microbiome in obese and lean twins. Nature 457, 480–484 (2009). https://doi.org/10.1038/nature07540

Within an individual oral cavity, we found over 3600 unique sequences, over 500 different OTUs or "species-level" phylotypes (sequences that clustered at 3% genetic difference) and 88 - 104 higher taxa (genus or more inclusive taxon). The predominant taxa belonged to Firmicutes (genus Streptococcus, family Veillonellaceae, genus Granulicatella), Proteobacteria (genus Neisseria, Haemophilus), Actinobacteria (genus Corynebacterium, Rothia, Actinomyces), Bacteroidetes (genus Prevotella, Capnocytophaga, Porphyromonas) and Fusobacteria (genus Fusobacterium).

ref:Zaura, E., Keijser, B.J., Huse, S.M. et al. Defining the healthy "core microbiome" of oral microbial communities. BMC Microbiol 9, 259 (2009). https://doi.org/10.1186/1471-2180-9-259

### Code and Figures:

```{r warning=F,fig.height=4,fig.width=5,dpi=600,fig.align='center',warning=F,out.width ="80%",results='hold'}
meta<-fread("../Data/Data/meta.csv",data.table = F)
OTUstool<-fread("../Data/Data/OTUtabale_regaStool.csv",data.table = F)
OTUsaliva<-fread("../Data/Data/OTUtabale_regaSaliva.csv",data.table = F)
TaxonLevels<-colnames(OTUstool)[1:7]
Stool<-OTUanalysis(OTUtable = OTUstool,
                    TaxonLevels = TaxonLevels,
                    topTaxonomyABvalue = c(0.01,0.01,0.1,0.2,0.2,0.2,0.2))

Saliva<-OTUanalysis(OTUtable = OTUsaliva,
                  TaxonLevels = TaxonLevels,
                  topTaxonomyABvalue = c(0.01,0.01,0.05,0.05,0.05,0.05,0.05))
                  
# Transform to compositional abundances
ABstool<-Stool$TaxonomyAbundance$Genus
ABsaliva<-Saliva$TaxonomyAbundance$Genus
# Pick the core (>5% relative abundance in >50% of the samples)
ABstool.core<-ABstool[apply(ABstool, 1, function(x){
  length(which(x==0))<34&length(which(x>0.05))!=0
}),]%>%
  .[-grep("norank",rownames(.)),]%>%
  .[-grep("unclassified",rownames(.)),]

mat1<-ABstool.core
for (i in 1:nrow(mat1)) {
  mat1[i,][which(mat1[i,]>=0.05)]=2
  mat1[i,][which(mat1[i,]>0&mat1[i,]<0.05)]=1
}

pheatmap(mat1,border_color = NA,
            width = unit(cm,"5"),height = unit(cm,"5"),
         fontsize = 6,
         main = "Prevalence of coreMicrobiome (Stool)",
         legend_labels = c("0","<0.05",">0.05"),
         legend_breaks = c(0,1,2),
         show_colnames = F,
         color = scico(100,palette = "bilbao",direction = 1))


ABsaliva.core<-ABsaliva[apply(ABsaliva, 1, function(x){
  length(which(x==0))<21&length(which(x>0.01))!=0
}),]%>%
  .[-grep("norank",rownames(.)),]%>%
  .[-grep("unclassified",rownames(.)),]

mat2<-ABsaliva.core
for (i in 1:nrow(mat2)) {
  mat2[i,][which(mat2[i,]>=0.05)]=2
  mat2[i,][which(mat2[i,]>0&mat2[i,]<0.05)]=1
}

pheatmap(mat2,border_color = NA,
         width = unit(cm,"5"),height = unit(cm,"5"),
         fontsize = 6,
         main = "Prevalence of coreMicrobiome (Saliva)",
         legend_labels = c("0","<0.05",">0.05"),
         legend_breaks = c(0,1,2),
         show_colnames = F,
         color = scico(100,palette = "bilbao",direction = 1))

ABstool.core<-data.frame(Samples=colnames(ABstool.core),t(ABstool.core))%>%
  mutate(.,Others=apply(.[,-1],1, function(x) 1-sum(x)))

ABsaliva.core<-data.frame(Samples=colnames(ABsaliva.core),t(ABsaliva.core))%>%
  mutate(.,Others=apply(.[,-1],1, function(x) 1-sum(x)))

coreM<-bind_rows(ABstool.core,ABsaliva.core)
coreM<-merge(select(meta,c(Samples,patientID,Site,Time)),coreM,by="Samples")
coreMp13<-subset(coreM,patientID=="Patient13")
coreMp13<-coreMp13[,-c(1:2)]%>%group_by(Site,Time)%>%summarise_each(mean)
coreMp13<-melt(coreMp13,id.vars = c("Site","Time"),value.name = "Abundance",
             variable.name = "CoreMicrobiome")
coreM1<-coreM[,-c(1:2)]%>%group_by(Site,Time)%>%summarise_each(mean)
coreM1<-melt(coreM1,id.vars = c("Site","Time"),value.name = "Abundance",
             variable.name = "CoreMicrobiome")
p1<-ggplot(coreM1,aes(Time,Abundance,fill=CoreMicrobiome))+
  geom_area()+
  facet_wrap(~Site)+
  theme_minimal()+
  scale_fill_manual(name="coreMicrobiome",values=c(col31,col21,col16,col11))+
  #scale_fill_discrete(guide=FALSE)+ #hide the legend
  theme(legend.text = element_text(size = 6),legend.box = "horizontal",
        plot.background = element_rect(colour = "black", 
                                       size = 1, linetype = 1,
                                       fill = "gray"),
        legend.margin = margin(0.1,unit="pt"),
        legend.key.size=unit(.1,"inches"),#图例分类符号的大小
        legend.text.align=0,#0左，1右，0.5居中， 图例分类标签的对齐方式
        legend.title=element_text(colour="black",size=8,face = "bold"),#图例标题设置
        legend.direction ="vertical",
        legend.box.just="top",
        legend.spacing = unit(0.1,"cm"),
        legend.spacing.y = unit(0.1,"cm"),
        legend.spacing.x =unit(0.1,"cm"),
        legend.box.spacing = unit(0.1,"cm"),
        legend.justification=c(.4,.4),
        legend.position="top",
        plot.margin = unit(c(0.2, 0.2, 0.2, 0.2), "inches"))+
  scale_x_continuous(n.breaks = 8,minor_breaks = NULL)+
  xlab("Cycles of treatment")
p1

p2<-ggplot(coreMp13,aes(Time,Abundance,fill=CoreMicrobiome))+
  geom_area()+
  facet_wrap(~Site,scales = "free")+
  theme_minimal()+
  scale_fill_manual(name="coreMicrobiome (Patient13)",values=c(col31,col21,col16,col11))+
  #scale_fill_discrete(guide=FALSE)+ #hide the legend
  theme(legend.text = element_text(size = 6),legend.box = "horizontal",
        plot.background = element_rect(colour = "black", 
                                       size = 1, linetype = 1,
                                       fill = "gray"),
        legend.margin = margin(0.1,unit="pt"),
        legend.key.size=unit(.1,"inches"),#图例分类符号的大小
        legend.text.align=0,#0左，1右，0.5居中， 图例分类标签的对齐方式
        legend.title=element_text(colour="black",size=8,face = "bold"),#图例标题设置
        legend.direction ="vertical",
        legend.box.just="top",
        legend.spacing = unit(0.1,"cm"),
        legend.spacing.y = unit(0.1,"cm"),
        legend.spacing.x =unit(0.1,"cm"),
        legend.box.spacing = unit(0.1,"cm"),
        legend.justification=c(.4,.4),
        legend.position="top",
        plot.margin = unit(c(0.2, 0.2, 0.2, 0.2), "inches"))+
  scale_x_continuous(n.breaks = 8,minor_breaks = NULL)+
  xlab("Cycles of treatment")
p2
```

## Microbiome compostion

### Code

```{r warning=F}
meta<-fread("../Data/Data/meta.csv",data.table = F)
OTUstool<-fread("../Data/Data/OTUtabale_regaStool.csv",data.table = F)
OTUsaliva<-fread("../Data/Data/OTUtabale_regaSaliva.csv",data.table = F)
TaxonLevels<-colnames(OTUstool)[1:7]
Stool<-OTUanalysis(OTUtable = OTUstool,
                    TaxonLevels = TaxonLevels,
                    topTaxonomyABvalue = c(0.01,0.01,0.1,0.2,0.2,0.2,0.2))

Saliva<-OTUanalysis(OTUtable = OTUsaliva,
                  TaxonLevels = TaxonLevels,
                  topTaxonomyABvalue = c(0.01,0.01,0.05,0.05,0.05,0.05,0.05))

stoolComp<-Stool$TaxonomyComposition%>%lapply(.,function(x){
  x<-mutate(x,Microname=rownames(x))
})
salivaComp<-Saliva$TaxonomyComposition%>%lapply(.,function(x){
  x<-mutate(x,Microname=rownames(x))
})

otu_abundance<-list()
areaPlot<-list()
for (i in seq_along(stoolComp)) {
  otu_abundance[[i]]<-merge(stoolComp[[i]],salivaComp[[i]],by="Microname",all = T)
  otu_abundance[[i]]<-reshape2::melt(otu_abundance[[i]],id.vars = "Microname",
                           variable.name = "Samples",value.name = "Abundance")
  otu_abundance[[i]]<-merge(meta,otu_abundance[[i]],by="Samples")
  otu_abundance[[i]]<-select(otu_abundance[[i]],c(Time,Site,Microname,Abundance))%>%
    group_by(Time,Site,Microname)%>%
    summarise_each(mean)
  otu_abundance[[i]]<-arrange(otu_abundance[[i]],desc(Abundance))
 ##plot
  areaPlot[[i]]<- ggplot(otu_abundance[[i]],aes(Time,Abundance,fill=Microname))+
    geom_area()+
    facet_wrap(~Site)+
    theme_minimal()+
    scale_fill_manual(name=names(stoolComp)[i],values=c(col31,col21,col16,col11))+
    theme(legend.text = element_text(size = 6),legend.box = "horizontal",
          plot.background = element_rect(colour = "black", 
                                         size = 1, linetype = 1,
                                         fill = "gray"),
          legend.margin = margin(0.1,unit="pt"),
          legend.key.size=unit(.1,"inches"),
          legend.text.align=0,
          legend.title=element_text(colour="black",size=8,face = "bold"),
          legend.direction ="vertical",
          legend.box.just="top",
          legend.spacing = unit(0.1,"cm"),
          legend.spacing.y = unit(0.1,"cm"),
          legend.spacing.x =unit(0.1,"cm"),
          legend.box.spacing = unit(0.1,"cm"),
          legend.justification=c(.4,.4),
          legend.position="top",
          plot.margin = unit(c(0.2, 0.2, 0.2, 0.2), "inches"))+
    scale_x_continuous(n.breaks = 8,minor_breaks = NULL)+
    xlab("Cycles of treatment")
  names(otu_abundance)[i]=names(stoolComp)[i]
  names(areaPlot)[i]=names(stoolComp)[i]
}
```

### Figures

<details>
<summary>
<font size=4>Figures</font>
</summary> 
#### Phylum compostion

```{r fig.height=5,fig.width=5,dpi=600,fig.align='center',warning=F,out.width ="100%",results='hold'}
print(areaPlot$Phylum)
```
#### Class compostion

```{r fig.height=5,fig.width=5,dpi=600,fig.align='center',warning=F,out.width ="100%",results='hold'}
print(areaPlot$Class)
```

#### Order compostion

```{r fig.height=5,fig.width=5,dpi=600,fig.align='center',warning=F,out.width ="100%",results='hold'}
print(areaPlot$Order)
```

#### Family compostion

```{r fig.height=5,fig.width=5,dpi=600,fig.align='center',warning=F,out.width ="100%",results='hold'}
print(areaPlot$Family)
```

#### Genus compostion

```{r fig.height=5,fig.width=6,dpi=600,fig.align='center',warning=F,out.width ="100%",results='hold'}
print(areaPlot$Genus)
```

#### Species compostion

```{r fig.height=5,fig.width=7,dpi=600,fig.align='center',warning=F,out.width ="100%",results='hold'}
print(areaPlot$Species)
```

#### OTU compostion

```{r fig.height=5,fig.width=6,dpi=600,fig.align='center',warning=F,out.width ="100%",results='hold'}
print(areaPlot$OTU)
```
</details>

